File: 2025/08 August/14 Count Reverse Pairs/main.cpp
// Link: https://www.geeksforgeeks.org/problems/count-reverse-pairs/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int merging(vector<int> &arr, int low, int mid, int high)
    {

        int count = 0, j = mid + 1;

        for (int i = low; i <= mid; i++)
        {
            while (j <= high && arr[i] > 2LL * arr[j])
            {
                j++;
            }
            count += (j - (mid + 1));
        }

        vector<int> temp;
        int left = low, right = mid + 1;

        while (left <= mid && right <= high)
        {
            if (arr[left] <= arr[right])
            {
                temp.push_back(arr[left++]);
            }
            else
            {
                temp.push_back(arr[right++]);
            }
        }

        while (left <= mid)
        {
            temp.push_back(arr[left++]);
        }

        while (right <= high)
        {
            temp.push_back(arr[right++]);
        }

        for (int i = low; i <= high; i++)
        {
            arr[i] = temp[i - low];
        }

        return count;
    }

    int mergeSort(vector<int> &arr, int low, int high)
    {

        if (low >= high)
        {
            return 0;
        }

        int mid = low + (high - low) / 2;
        int count = mergeSort(arr, low, mid) +
                    mergeSort(arr, mid + 1, high) +
                    merging(arr, low, mid, high);

        return count;
    }

    int countRevPairs(vector<int> &arr)
    {
        return mergeSort(arr, 0, arr.size() - 1);
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {3, 2, 4, 5, 1, 20};
    int result = sol.countRevPairs(arr);
    cout << "Count of reverse pairs: " << result << endl;

    arr = {5, 4, 3, 2, 2};
    result = sol.countRevPairs(arr);
    cout << "Count of reverse pairs: " << result << endl;
    return 0;
}

File: 2025/08 August/14 Count Reverse Pairs/main.py
# Link: https://www.geeksforgeeks.org/problems/count-reverse-pairs/1


class Solution:
    def merging(self, arr, low, mid, high):
        count = 0
        j = mid + 1

        for i in range(low, mid + 1):
            while j <= high and arr[i] > 2 * arr[j]:
                j += 1
            count += j - (mid + 1)

        temp = []
        left, right = low, mid + 1

        while left <= mid and right <= high:
            if arr[left] <= arr[right]:
                temp.append(arr[left])
                left += 1
            else:
                temp.append(arr[right])
                right += 1

        while left <= mid:
            temp.append(arr[left])
            left += 1

        while right <= high:
            temp.append(arr[right])
            right += 1

        for i in range(low, high + 1):
            arr[i] = temp[i - low]

        return count

    def mergeSort(self, arr, low, high):

        if low >= high:
            return 0

        mid = low + (high - low) // 2
        count = (
            self.mergeSort(arr, low, mid)
            + self.mergeSort(arr, mid + 1, high)
            + self.merging(arr, low, mid, high)
        )

        return count

    def countRevPairs(self, arr):
        return self.mergeSort(arr, 0, len(arr) - 1)


if __name__ == "__main__":
    sol = Solution()
    arr = [3, 2, 4, 5, 1, 20]
    result = sol.countRevPairs(arr)
    print(f"Count of reverse pairs: {result}")

    arr = [5, 4, 3, 2, 2]
    result = sol.countRevPairs(arr)
    print(f"Count of reverse pairs: {result}")
